[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73703

This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

Commit:
[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op

This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/math_ops.py:
```diff
@@ -0,0 +1,75 @@
+            else:
+                raise RuntimeError("Implicit broadcasting not supported yet!")
+
+        elif isinstance(lhs, ReplicatedTensor):
+            assert isinstance(rhs, ShardedTensor)
+            if lhs.size() == rhs.local_tensor().size():
+                res = op(lhs, rhs.local_tensor())
+                return ShardedTensor._init_from_local_tensor(res, rhs.sharding_spec(), rhs.size(), process_group=pg)
+            else:
+                raise RuntimeError("Implicit broadcasting not supported yet!")
+
+        elif isinstance(rhs, ReplicatedTensor):
+            assert isinstance(lhs, ShardedTensor)
+            if rhs.size() == lhs.local_tensor().size():
+                res = op(lhs.local_tensor(), rhs)
+                return ShardedTensor._init_from_local_tensor(res, lhs.sharding_spec(), lhs.size(), process_group=pg)
+            else:
+                raise RuntimeError("Implicit broadcasting not supported yet!")
```
Not a blocker here. But do you feel we can make consolidate the logic a little bit? They are essentially the same right?

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/math_ops.py:
```diff
@@ -0,0 +1,75 @@
+    Tensor.sub,
+    Tensor.__rsub__,
+    # mul
+    torch.mul,
+    Tensor.mul,
```
Thanks for making this happens. Can you also kindly add `torch.baddbmm` and `torch.bmm` because Megatron-LM use it for Transformer model (https://github.com/NVIDIA/Megatron-LM/blob/e156d2fea7fc5c98e645f7742eb86b643956d840/megatron/model/transformer.py#L349). So I guess instead of creating a new PR, adding it here might be faster? If you feel this PR is too big, I can work on another PR on top of this one. Thanks!

wanchaol replied:
sure, let me try adding it in the PR

pritamdamania87 replied:
@fduwjj @wanchaol Let us focus on the fairseq XLMR-XL ops first: https://github.com/pytorch/fairseq/blob/c71870f370455e6154c730e8822ea323b5f266f6/fairseq/modules/transformer_layer.py#L20. Also, note that `torch.baddbmm` and `torch.bmm` are not binary math ops. But rather matmul ops.

Lets address these in a separate PR

wanchaol replied:
yeah sounds good, I actually tried with `baddbmm` and `bmm`, it was a bit complicated as it involves communication (similar to linear). As pritam suggested, given that they are matmul ops and not binary ops, lets leave it out of this PR and we can address the XLMR-XL ops first @fduwjj

fduwjj replied:
Sure. But it looks like `bmm` is indeed used in the XLMR: https://github.com/pytorch/fairseq/blob/c71870f370455e6154c730e8822ea323b5f266f6/fairseq/modules/multihead_attention.py#L504. I think I can work on this on my end.

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/math_ops.py:
```diff
@@ -0,0 +1,75 @@
+        lhs = args[0]
+        rhs = args[1]
+        # Validate types
+        if isinstance(lhs, ShardedTensor) and isinstance(rhs, ShardedTensor):
+            lhs_spec = lhs.sharding_spec()
+            rhs_spec = rhs.sharding_spec()
+            if not isinstance(lhs_spec, ChunkShardingSpec) or not isinstance(rhs_spec, ChunkShardingSpec):
+                raise TypeError("Only ShardedTensor with ChunkShardingSpec supports"
+                                " two ShardedTensor together")
+
+            if lhs.size() == rhs.size() and lhs_spec.sharding_dim == rhs_spec.sharding_dim:
+                # perform local element-wise math op
+                res = op(lhs.local_tensor(), rhs.local_tensor)
+                return ShardedTensor._init_from_local_tensor(res, lhs_spec, lhs.size(), process_group=pg)
+            else:
+                raise RuntimeError("Implicit broadcasting not supported yet!")
```
Nice! Can you also add test cases for sharded tensor + sharded tensor interations?

fduwjj reviewed (commented):
Overall, this looks great!! Thanks for doing that and I just have some minor nits here. If you need to land this faster, I can unblock it before the end of this week.

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/_ops/math_ops.py:
```diff
@@ -0,0 +1,75 @@
+
+        elif isinstance(lhs, ReplicatedTensor):
+            assert isinstance(rhs, ShardedTensor)
+            if lhs.size() == rhs.local_tensor().size():
```
Why do we need this check? Wouldn't `res = op(lhs, rhs.local_tensor())` work even if this is not True? Broadcasting within a shard should be equivalent to broadcasting across the whole tensor since we are applying the same ReplicatedTensor on all ranks?

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/_ops/math_ops.py:
```diff
@@ -0,0 +1,75 @@
+
+        elif isinstance(rhs, ReplicatedTensor):
+            assert isinstance(lhs, ShardedTensor)
+            if rhs.size() == lhs.local_tensor().size():
```
same as above.

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -790,6 +793,24 @@ def __hash__(self):
     def __repr__(self):
         return f'ShardedTensor({self._metadata})'
 
+    def __add__(self, other):
+        return handle_torch_function(torch.Tensor.__add__, (self, other), self, other)
+
+    def __sub__(self, other):
+        return handle_torch_function(torch.Tensor.__sub__, (self, other), self, other)
+
+    def __rsub__(self, other):
+        return handle_torch_function(torch.Tensor.__rsub__, (self, other), self, other)
+
+    def __mul__(self, other):
+        return handle_torch_function(torch.Tensor.__mul__, (self, other), self, other)
+
+    def __div__(self, other):
+        return handle_torch_function(torch.Tensor.__div__, (self, other), self, other)
+
+    def __rdiv__(self, other):
+        return handle_torch_function(torch.Tensor.__rdiv__, (self, other), self, other)
```
Is this the recommended pattern to support things like `tensor.add` which I guess won't be covered by __torch_function__?

wanchaol replied:
Ideally, things like this should be handled by `__torch_function__`, the reason that here does not getting handled automatically, is that `ShardedTensor` is not a subclass of torch.Tensor, so I think once https://github.com/pytorch/pytorch/pull/74695 landed, we are able to remove these magic methods

pritamdamania87 commented on test/distributed/_shard/sharded_tensor/ops/test_math_ops.py:
```diff
@@ -0,0 +1,130 @@
+        import builtins
+        ops = ["torch.add", "torch.sub", "torch.mul", "torch.div", "+", "-", "*"]
+
+        def gen_code(python_op):
```
This is cool haha :)

pritamdamania87 commented on test/distributed/_shard/sharded_tensor/ops/test_math_ops.py:
```diff
@@ -0,0 +1,130 @@
+
+                self.assertEqual(output, global_output)
+
+            # test basic math ops between ShardedTensor and scalar
```
should we test float too?

pritamdamania87 commented on test/distributed/_shard/test_replicated_tensor.py:
```diff
@@ -74,3 +76,109 @@ def test_replicated_tensor_inter_op_tensor(self):
+            isinstance(new_tensor, sharded_tensor.ShardedTensor)
+            and not isinstance(new_tensor, ReplicatedTensor)
+        )
+        self.assertEqual(new_tensor.local_tensor(), st.local_tensor() + 4)
+        # test torch.add
+        new_tensor = torch.add(replica_tensor, st)
+        self.assertEqual(new_tensor.local_tensor(), st.local_tensor() + 4)
+
+        new_tensor = st + replica_tensor
+        self.assertEqual(new_tensor.local_tensor(), st.local_tensor() + 4)
+
+        # sub op
+        new_tensor = replica_tensor - st
+        self.assertTrue(
```
Shouldn't this move to `test_math_ops`?

pritamdamania87 commented on test/distributed/_shard/test_replicated_tensor.py:
```diff
@@ -74,3 +76,109 @@ def test_replicated_tensor_inter_op_tensor(self):
+    @skip_if_lt_x_gpu(4)
+    @requires_nccl()
+    def test_replicated_tensor_inter_op_sharded_tensor(self):
+        local_tensor = torch.ones(3, 3, device=f"cuda:{self.rank}") * 4
+        replica_tensor = ReplicatedTensor(local_tensor)
```
So I think we should do the following to ensure our assumptions around broadcasting with ReplicatedTensor are valid:

1. Create two local random tensors (t1, t2) and run all of these ops on them.
2. Now create a ShardedTensor from t1 and ReplicatedTensor from t2 and run the same ops.
3. Gather the results (ex: on rank 0) and validate to a local ops.

pritamdamania87 commented on test/distributed/_shard/test_replicated_tensor.py:
```diff
@@ -74,3 +76,109 @@ def test_replicated_tensor_inter_op_tensor(self):
+
+        st = sharded_tensor.rand(spec, (20, 3))
+
+        with self.assertRaisesRegex(RuntimeError, 'Implicit broadcasting'):
+            st + replica_tensor
```
This should work with the latest changes in the PR right?

pritamdamania87 commented on test/distributed/_shard/test_replicated_tensor.py:
```diff
@@ -74,3 +76,109 @@ def test_replicated_tensor_inter_op_tensor(self):
+            st + replica_tensor
+
+        with self.assertRaisesRegex(TypeError, 'unsupported operand type'):
+            st + 4
```
Why doesn't this work? I thought we enabled it by handling scalars in the binary ops?

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_shard/replicated_tensor.py:
```diff
@@ -73,11 +73,15 @@ def dispatch_arg(arg):
 
         for arg in args:
-            dispatch_arg(arg)
+            redispatched_res = dispatch_arg(arg)
+            if redispatched_res:
+                return redispatched_res
```
I'm not sure if this would work as expected always. For example, for things like in-place ops there is no return value. In such cases, don't we still erroneously go to kwargs?

Maybe `dispatch_arg` should return a tuple of return value and whether the function was executed or not.

wanchaol replied:
addressed.

pritamdamania87 commented on test/distributed/_shard/test_replicated_tensor.py:
```diff
@@ -74,3 +77,143 @@ def test_replicated_tensor_inter_op_tensor(self):
+        output = torch.empty((12, 3)) if self.rank == 0 else None
+        res.gather(dst=0, out=output)
+
+        if self.rank == 0:
+            local_output = torch.add(local_tensor1, local_tensor2)
+            self.assertEqual(output, local_output)
```
It looks like we are doing this local vs sharded + replicated verification only here for torch.add. Shouldn't we run `test_basic_math_ops` with ShardedTensor op ReplicatedTensor and verify it is on par with local compute?

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/_ops/_common.py:
```diff
@@ -378,3 +378,23 @@ def _handle_max_norm_col_wise(
     return local_shard_norm_renormed, gathered_inputs
+
+
+def narrow_tensor(tensor, metadata):
+    """
+    narrow the tensor according to the metadata
+    """
+    narrowed_tensor = tensor
+    shard_offsets = metadata.shard_offsets
+    shard_sizes = metadata.shard_sizes
+    for idx, (offset, size) in enumerate(zip(shard_offsets, shard_sizes)):
+        if size < tensor.size(idx):
+            # Reshape to get shard for this rank and we don't want autograd
+            # recording here for the narrow op and 'local_shard' should be a
+            # leaf variable in the autograd graph.
+            narrowed_tensor = narrowed_tensor.narrow(
+                idx,
+                shard_offsets[idx],
+                shard_sizes[idx]
+            )
+    return narrowed_tensor
```
Don't we do something similar in `_shard_parameter`/`_shard_tensor`? Maybe we can consolidate the code to use this method?

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/_ops/math_ops.py:
```diff
@@ -49,7 +61,14 @@ def binary_math_op(types, args=(), kwargs=None, pg=None):
+            st_meta = lhs.local_shards()[0].metadata
+            if st_size != lhs.size():
+                # try to broadcast replicated tensor
+                rhs.expand(st_size)
```
Wondering why `test_replicated_tensor_inter_op_sharded_tensor` didn't catch this issue? Or am I missing something here?

wanchaol replied:
just fixed this, the `test_replicated_tensor_inter_op_sharded_tensor` was not triggering bc it does not trigger implicit broadcasting, I fixed the `test_replicated_tensor_implicit_broadcasting` and use size of (3,) for replicatedtensor, locally verified it triggers the failure if we don't use the return value of expand.

pritamdamania87 replied:
Hmm, so why doesn't (1, 3) trigger broadcasting as well? It should right?

wanchaol replied:
For the case of (1, 3), it right now should correctly trigger the `expand` logic as well. The reason it was not triggered is because:
* There were some logic wrong to trigger the `expand` call, I used the wrong operands to compare the size
* even if it was not getting expanded, `narrow_tensor` call still returns the original tensor instead of throwing as it's too small to narrow on dim 0
* so `res = op(lhs.local_tensor(), replica_part)` got executed with size of `op((3, 3), (1, 3))`, where the implicit broadcasting happened in local shard computation

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

wanchaol commented on test/distributed/_shard/test_replicated_tensor.py:
```diff
@@ -82,10 +85,11 @@ def test_replicated_tensor_inter_op_tensor(self):
         torch.manual_seed(self.rank)
+
+        local_tensor1 = torch.rand(12, 3, device=f"cuda:{self.rank}") * 4
+        local_tensor2 = torch.ones(12, 3, device=f"cuda:{self.rank}") * 4
```
So this is because we need to compare local gathered results and sharded results with the same ReplicatedTensor content, if it's `torch.rand`, the replicated tensor content on each rank will be different so the gathered results will be different and not comparable with the local result

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]

Commit:
Update on "[shard] Add basic math ops to ShardedTensor and add ReplicatedTensor inter-op"


This PR add basic math ops to ShardedTensor (+-*/), and add ReplicatedTensor inter-op ShardedTensor to those math ops. This enables ShardedTensor (op) ReplicatedTensor to avoid communication in certain cases.

Differential Revision: [D34560867](https://our.internmc.facebook.com/intern/diff/D34560867/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34560867/)!

[ghstack-poisoned]
